src/dbus_api.rs

```diff
@@ -0,0 +1,328 @@
+}
+
+impl DBusApi {
+    pub fn new(base: &'static str, method_retry_error_names: Vec<&'static str>) -> Self {
+        let connection = DBusConnection::get_private(BusType::System).unwrap();
+
+        DBusApi {
+            connection: connection,
+            method_timeout: DEFAULT_TIMEOUT,
+            base: base,
+            method_retry_error_names: method_retry_error_names,
+        }
+    }
+
+    pub fn method_timeout(&self) -> u64 {
+        self.method_timeout
+    }
+
+    pub fn set_method_timeout(&mut self, timeout: u64) {
+        self.method_timeout = timeout;
+    }
+
+    pub fn call(&self, path: &str, interface: &str, method: &str) -> Result<Message, String> {
+        self.call_with_args(path, interface, method, vec![])
+    }
+
+    pub fn call_with_args(&self,
+                          path: &str,
+                          interface: &str,
+                          method: &str,
+                          args: Vec<&RefArg>)
+                          -> Result<Message, String> {
+        let call_error = |details: &str| {
+            Err(format!("D-Bus '{}'::'{}' method call failed on '{}': {}",
+                        interface,
+                        method,
+                        path,
+                        details))
+        };
+
+        match self.call_with_args_retry(path, interface, method, args) {
+            Ok(response) => Ok(response),
+            Err(error) => call_error(&error),
+        }
+    }
+
+    fn call_with_args_retry(&self,
+                            path: &str,
+                            interface: &str,
+                            method: &str,
+                            args: Vec<&RefArg>)
+                            -> Result<Message, String> {
+        let mut retries = 0;
+
+        loop {
+            if let Ok(result) = self.create_and_send_message(path, interface, method, &args) {
+                return result;
+            }
+
+            retries += 1;
+
+            if retries == RETRIES_ALLOWED {
+                return Err(format!("method failed after {} retries", RETRIES_ALLOWED));
+            }
+
+            ::std::thread::sleep(::std::time::Duration::from_secs(1));
+        }
+    }
+
+    fn create_and_send_message(&self,
+                               path: &str,
+                               interface: &str,
+                               method: &str,
+                               args: &Vec<&RefArg>)
+                               -> Result<Result<Message, String>, String> {
+        match Message::new_method_call(self.base, path, interface, method) {
+            Ok(mut message) => {
+                if args.len() > 0 {
+                    message = message.append_ref(args);
+                }
+
+                self.send_message_checked(message)
+            }
+            Err(details) => Ok(Err(details)),
+        }
+    }
+
+    fn send_message_checked(&self, message: Message) -> Result<Result<Message, String>, String> {
```

Comment:
Nice catch, all I actually needed is a `Result` wrapped in an `Option`. Done in 10fa8ca